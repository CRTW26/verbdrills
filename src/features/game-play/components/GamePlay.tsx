import React, { useEffect, useState } from 'react'
import Button from 'shared/components/Button'
import Input from 'shared/components/Input'
import Timer from './Timer'

interface Verb {
  infinitive: string
  person: string
  translation: string
  answer: string
}

interface Props {
  verb: Verb
  score: number
  validateInput: (guess: string, answer: string) => void
  onExpiration: () => void
}

const GamePlay: React.FC<Props> = ({
  verb,
  score,
  validateInput,
  onExpiration,
}) => {
  const [guess, setGuess] = useState('')

  const { infinitive, person, translation, answer } = verb

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setGuess(e.target.value)
  }

  const submitGuess = () => {
    validateInput(guess, answer)
  }

  const handleEnterPress = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      submitGuess()
    }
  }

  useEffect(() => {
    setGuess('')
  }, [score])

  // TODO: look into useEffect taking snapshot of state. This uses initial state
  // useEffect(() => {
  //   window.addEventListener('keypress', handleEnterPress)

  //   return () => {
  //     window.removeEventListener('keypress', handleEnterPress)
  //   }
  // }, [])

  return (
    <div className="game-play">
      <Timer onExpiration={onExpiration} />

      <div>
        <h2>{score}</h2>
      </div>

      <div className="card__child card__verb-display">
        <h2 className="infinitive">{infinitive}</h2>
        <h2 className="person">{person}</h2>
        <h3 className="translation">{translation}</h3>
      </div>

      <div className="card__child card__input">
        <Input
          value={guess}
          onChange={handleChange}
          onKeyDown={handleEnterPress}
        />

        <Button
          className="btn btn--primary"
          text="Guess"
          onClick={submitGuess}
        />
      </div>

      <style jsx>{`
        .game-play {
          width: 100%;
          text-align: center;
        }
      `}</style>
    </div>
  )
}

export default GamePlay
