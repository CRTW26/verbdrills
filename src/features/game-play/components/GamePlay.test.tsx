import { render } from '@testing-library/react'
import GamePlay from './GamePlay'

describe('<GamePlay />', () => {
  test('should render', () => {
    const verb = {
      infinitive: 'hablar',
      person: 'yo',
      translation: 'to speak',
      answer: 'acepto',
    }

    const { asFragment } = render(
      <GamePlay
        verb={verb}
        score={0}
        validateInput={jest.fn()}
        onExpiration={jest.fn()}
      />
    )

    expect(asFragment).toMatchSnapshot()
  })
})
