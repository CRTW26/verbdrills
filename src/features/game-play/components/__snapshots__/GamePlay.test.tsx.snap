// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GamePlay /> should render 1`] = `
<DocumentFragment>
  <div
    class="jsx-2177346083 game-play"
  >
    <div>
      <h3>
        0:6
      </h3>
    </div>
    <div
      class="jsx-2177346083"
    >
      <h2
        class="jsx-2177346083"
      >
        0
      </h2>
    </div>
    <div
      class="jsx-2177346083 card__child card__verb-display"
    >
      <h2
        class="jsx-2177346083 infinitive"
      >
        hablar
      </h2>
      <h2
        class="jsx-2177346083 person"
      >
        yo
      </h2>
      <h3
        class="jsx-2177346083 translation"
      >
        to speak
      </h3>
    </div>
    <div
      class="jsx-2177346083 card__child card__input"
    >
      <input
        class="jsx-828487101 "
        value=""
      />
      <button
        class="jsx-347471370 btn btn--primary"
      >
        Guess
      </button>
    </div>
  </div>
</DocumentFragment>
`;
